core/src/main/java/org/eclipse/hono/util/TenantObject.java

```diff
@@ -117,112 +131,90 @@ public TenantObject setEnabled(final boolean flag) {
-    public TenantObject setTrustAnchor(final X509Certificate certificate) {
+    public TenantObject addTrustAnchor(final X509Certificate certificate) {
 
         Objects.requireNonNull(certificate);
 
-        try {
-            final JsonObject trustedCa = new JsonObject()
-                    .put(TenantConstants.FIELD_PAYLOAD_CERT, certificate.getEncoded());
-            return setProperty(TenantConstants.FIELD_PAYLOAD_TRUSTED_CA, trustedCa);
-        } catch (final CertificateEncodingException e) {
-            throw new IllegalArgumentException("cannot encode certificate");
-        }
-    }
-
-    /**
-     * Gets the trusted certificate authority configured for this tenant.
-     * <p>
-     * This method tries to extract the certificate from the data contained in
-     * the JSON object of the <em>trusted-ca</em> property.
-     * The value of the JSON object's <em>cert</em> property is expected to contain
-     * the Base64 encoded <em>binary</em> DER-encoding of the certificate, i.e. the same
-     * value as the <em>printable</em> form but without the leading
-     * {@code -----BEGIN CERTIFICATE-----} and trailing {@code -----END CERTIFICATE-----}.
-     * 
-     * @return The certificate or {@code null} if no certificate authority
-     *         has been set.
-     * @throws CertificateException if the value of <em>cert</em> property cannot be parsed into
-     *             an X.509 certificate.
-     */
-    @JsonIgnore
-    public X509Certificate getTrustedCertificateAuthority() throws CertificateException {
-
-        final JsonObject trustedCa = getProperty(TenantConstants.FIELD_PAYLOAD_TRUSTED_CA, JsonObject.class);
-        if (trustedCa == null) {
-            return null;
+        final String subjectDn = certificate.getSubjectX500Principal().getName();
+        if (Strings.isNullOrEmpty(subjectDn)) {
+            throw new IllegalArgumentException("subject DN of certificate is either missing or empty");
         } else {
-            final Object obj = trustedCa.getValue(TenantConstants.FIELD_PAYLOAD_CERT);
-            if (obj instanceof String) {
-                try {
-                    final byte[] derEncodedCert = Base64.getDecoder().decode(((String) obj));
-                    final CertificateFactory factory = CertificateFactory.getInstance("X.509");
-                    return (X509Certificate) factory.generateCertificate(new ByteArrayInputStream(derEncodedCert));
-                } catch (final IllegalArgumentException e) {
-                    // Base64 decoding failed
-                    throw new CertificateParsingException("cert property does not contain valid Base64 scheme", e);
-                }
-            } else {
-                return null;
+            try {
+                final JsonObject trustedCa = new JsonObject()
+                        .put(TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, subjectDn)
+                        .put(TenantConstants.FIELD_PAYLOAD_CERT, certificate.getEncoded());
```

Comment:
IMHO we should *always* only store the key, subject DN and validity period, and *not* the certificate itself. Accepting a certificate to add to the trust store is just a convenience but the internal data structure should always be the same. This will spare us many checks when processing a TenantObject instance ...

Reply:
true 👍